### PR TITLE
 test(wasm-gen): cover all the precise sys-calls with tests 

### LINKED ISSUE #3171: Cover all the precise sys-calls in `wasm-gen` with tests
### Problem to Solve

There's some sys-calls, called `Precise` that actually consist of many syscalls in `wasm-gen` (#3091). They aren't tested anywhere for now

### Possible Solution

 Cover all the `Precise` syscalls with tests

### Notes

This is the tracking issue for these precise syscalls:
- [x] `reply_from_reservation`
- [x] `send_init` -> `send_push`/`send_push_input` -> `send_commit`

### PR DESCRIPTION
**Pull request by StackOverflowExcept1on**
Resolves #3171

also added `env_logger` to sys-calls tests:

```
RUST_LOG="syscalls=trace" cargo test --package gear-wasm-gen --release --lib -- tests::precise_syscalls_works --exact --nocapture
```

### CODE REVIEW

File: utils/wasm-gen/src/tests.rs [hunk 1]
```diff
@@ -199,8 +226,69 @@ fn error_processing_works_for_fallible_syscalls() {
...
+            termination_reason,
+            TerminationReason::Actor(ActorTerminationReason::Success),
+            "syscall: {}",
+            syscall.to_str()
+        );
+    }
+}
+
+#[test]
+fn precise_syscalls_works() {
```
Comment (mertwole) [inline utils/wasm-gen/src/tests.rs:248]: I think this logics can be shared with the one used in fn above, counting what I've written in the other review comments

**Review by mertwole — CHANGES_REQUESTED** [review]

File: utils/wasm-gen/src/tests.rs [hunk 2]
```diff
@@ -234,50 +326,27 @@ fn get_params_for_syscall_to_fail(
     )
 }
 
-fn execute_wasm_with_syscall_injected(
-    syscall: SysCallName,
-    ignore_fallible_errors: bool,
-    params_config: SysCallsParamsConfig,
+fn execute_wasm_with_custom_configs(
+    unstructured: &mut Unstructured,
```
[the remark this answers is not in the document]
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/tests.rs:330]: Make sense, I'll create config `number_of_pushes` in `SysCallsConfig`

File: utils/wasm-gen/src/tests.rs [hunk 3]
```diff
@@ -234,50 +326,27 @@ fn get_params_for_syscall_to_fail(
...
+    sys_calls_config: SysCallsConfig,
     initial_memory_write: Option<MemoryWrite>,
+    initial_pages: u16,
+    context_settings: ContextSettings,
```
[the remark this answers is not in the document]
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/tests.rs]: done in e730c8fbed1e58a2b6cdbbd1a5a7b707ac0da9c3

File: utils/wasm-gen/src/tests.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/tests.rs:322]: done in 32deba85ae026715048c26de8994d1c066f10f47

File: utils/wasm-gen/src/config/syscalls/precise.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum PreciseSysCallAdditionalData {
+    Nothing,
+    Range(RangeInclusive<usize>),
+}
+
+/// Possible additional data for each precise sys-call.
+/// Can be used to write unit tests so you don't have to rely on randomness.
+#[derive(Debug, Clone)]
+pub struct SysCallsPreciseConfig(HashMap<SysCallName, PreciseSysCallAdditionalData>);
```
Comment (mertwole) [inline utils/wasm-gen/src/config/syscalls/precise.rs:34]: Why do we need to generalize it? You can just introduce concrete config like `number_of_send_push_calls_generated: RangeInclusive<usize>`

File: utils/wasm-gen/src/tests.rs [hunk 5]
```diff
@@ -176,19 +176,42 @@ fn injecting_addresses_works() {
 fn error_processing_works_for_fallible_syscalls() {
     use gear_backend_common::ActorTerminationReason;
 
+    let _ = env_logger::builder().is_test(true).try_init();
+
+    // We create Unstructured from zeroes here as we just need any
```
Comment (mertwole) [inline utils/wasm-gen/src/tests.rs:181]: It was my comment, but anyway)
```suggestion
    // We create Unstructured from zeroes here as we just need any.
```

**Review by mertwole — CHANGES_REQUESTED** [review]

File: utils/wasm-gen/src/config/syscalls/precise.rs (hunk 4 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config/syscalls/precise.rs:34]: This is something too specific. What if we have some other pattern and we need other data?

File: utils/wasm-gen/src/tests.rs [hunk 6]
```diff
@@ -199,8 +222,65 @@ fn error_processing_works_for_fallible_syscalls() {
...
+fn precise_syscalls_works() {
+    use gear_backend_common::ActorTerminationReason;
+
+    let _ = env_logger::builder().is_test(true).try_init();
+
+    // Pin a specific seed for this test.
+    let mut rng = SmallRng::seed_from_u64(1234);
+    let mut buf = vec![0; UNSTRUCTURED_SIZE];
+    rng.fill_bytes(&mut buf);
+    let mut unstructured = Unstructured::new(&buf);
```
[the remark this answers is not in the document]
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/tests.rs:252]: Completely removing `Unstructured` as in the first test will result in pointers that will always be `0x00`

File: utils/wasm-gen/src/config/syscalls/precise.rs (hunk 4 above)
Comment (StackOverflowExcept1on) [reply utils/wasm-gen/src/config/syscalls/precise.rs]: done in b2dc1cdf29e0bbfb997d370db708a2b73e4b5595

File: utils/wasm-gen/src/config/syscalls/precise.rs [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Entities describing configuration for precise sys-calls.
+
+use std::ops::RangeInclusive;
+
+/// Represents the configuration for precise sys-calls.
```
Comment (SabaunT) [inline utils/wasm-gen/src/config/syscalls/precise.rs:23]:
```suggestion
/// Represents the configuration for building some parts of precise sys-calls.
```

File: utils/wasm-gen/Cargo.toml [hunk 8]
```diff
@@ -23,6 +23,7 @@ rand = { workspace = true, features = ["small_rng"] }
 wat.workspace = true
 indicatif.workspace = true
 proptest.workspace = true
+env_logger.workspace = true
```
Comment (SabaunT) [inline utils/wasm-gen/Cargo.toml:26]: we use logger from `gear_utils`

**Review by SabaunT — APPROVED** [review]

**Review by breathx — APPROVED** [review]

**Comment by breathx** [comment]
@mertwole re review please

**Review by mertwole — APPROVED** [review]

**Pull request merged**